sdk/go/pulumi/context.go

```diff
@@ -747,34 +740,73 @@ func (ctx *Context) InvokePackage(
 			ferr = multierror.Append(ferr,
 				fmt.Errorf("%s invoke failed: %s (%s)", tok, failure.Reason, failure.Property))
 		}
-		return ferr
+		return nil, ferr
 	}
 
 	// Otherwise, simply unmarshal the output properties and return the result.
-	outProps, err := plugin.UnmarshalProperties(
+	return plugin.UnmarshalProperties(
 		resp.Return,
 		plugin.MarshalOptions{
 			KeepUnknowns:  true,
 			KeepSecrets:   true,
 			KeepResources: true,
-		},
-	)
-	if err != nil {
-		return err
+		})
+}
+
+func validInvokeResult(resultV reflect.Value) bool {
+	isPointer := resultV.Kind() == reflect.Ptr
+	isMap := resultV.Elem().Kind() == reflect.Map && resultV.Elem().Type().Key().Kind() == reflect.String
+	structOrMap := resultV.Elem().Kind() == reflect.Struct || isMap
+	return isPointer && structOrMap
+}
+
+// InvokePackage will invoke a provider's function, identified by its token tok. This function call is synchronous.
+//
+// args and result must be pointers to struct values fields and appropriately tagged and typed for use with Pulumi.
+func (ctx *Context) InvokePackage(
+	tok string, args interface{}, result interface{}, packageRef string, opts ...InvokeOption,
+) error {
+	resultV := reflect.ValueOf(result)
+	if !validInvokeResult(resultV) {
+		return errors.New("result must be a pointer to a struct or map value")
 	}
 
-	// fail if there are secrets returned from the invoke
-	hasSecret, err := unmarshalOutput(ctx, resource.NewObjectProperty(outProps), resultV.Elem())
+	outProps, err := ctx.invokePackageRaw(tok, args, packageRef, opts...)
 	if err != nil {
 		return err
 	}
+	hasSecret, err := unmarshalOutput(ctx, resource.NewObjectProperty(outProps), resultV.Elem())
+
 	if hasSecret {
-		return errors.New("unexpected secret result returned to invoke call")
+		return errors.New("unexpected secret result returned to invoke call, " +
+			"consider using the output-versioned variant of the invoke")
```

Comment:
Changed the error message here slightly to indicate that users should be using the output-versioned invoke